### DISCUSSION TITLE
gradle Wrong offset errors

### DISCUSSION THREAD

**Discussion by japj** [renovatebot/renovate#12399] [Request Help] [answered]
Hello,

I have some weird issues with gradle, where failures occur with "Wrong offset errors".

This is with a selfhosted renovate (28.11.2) connected to an AzureDevOps Server git repository.

The renovate.json contains:
```json
  "extends": [
    "config:base",
    "group:allNonMajor",
    ":preserveSemverRanges"
  ]
```

The build.gradle contains:
```gradle
plugins {
    id 'java'
    id 'application'
    id 'maven-publish'
}


dependencies {
    implementation 'com.google.guava:guava:28.0-jre'
    testImplementation 'junit:junit:4.12'
    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.4.2'
}
```

The error that I get is:
```txt
2021-10-29T14:29:53.9298400Z   DEBUG: Wrong offset (repository=some/repository, branch=renovate/all-minor-patch)
2021-10-29T14:29:53.9299111Z          "depName": "org.junit.jupiter:junit-jupiter-engine",
2021-10-29T14:29:53.9299622Z          "currentValue": "5.4.2",
2021-10-29T14:29:53.9300009Z          "newValue": "5.8.1"
```

It seems from looking at https://github.com/renovatebot/renovate/blob/569bee6e7d78956bac61831794ba54272f314e42/lib/manager/gradle/shallow/update.ts#L26 it is logging the **original** currentValue, not the **actual** currentValue, which makes it a bit tricky to analyze.

I think this issue is due to applying multiple updates to the same gradle file.
In this specific case it goes 'wrong' when one of these updates changes the 'version string length',  causing a subsequent `fileReplacePosition` to not be correct anymore.

The below baseDeps shows that the `junit:junit` version will go from `4.12` to `4.13.2` (at filePosition 185)
This will cause the version string for `junit-jupiter-engine` (at filePosition 252) to have moved, causing the "Wrong offset" error.

```json
  "baseDeps": [
    {
      "depName": "com.google.guava:guava",
      "currentValue": "28.0-jre",
      "managerData": {
        "fileReplacePosition": 138,
        "packageFile": "Test.Build.Gradle.Java/build.gradle"
      },
      "fileReplacePosition": 138,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2"
      ],
      "depIndex": 0,
      "updates": [
        {
          "bucket": "non-major",
          "newVersion": "28.2-jre",
          "newValue": "28.2-jre",
          "newMajor": 28,
          "newMinor": 2,
          "updateType": "minor",
          "branchName": "renovate/all-minor-patch"
        },
        {
          "bucket": "major",
          "newVersion": "31.0.1-jre",
          "newValue": "31.0.1-jre",
          "newMajor": 31,
          "newMinor": 0,
          "updateType": "major",
          "branchName": "renovate/com.google.guava-guava-31.x"
        }
      ],
      "warnings": [],
      "versioning": "gradle",
      "sourceUrl": "https://github.com/google/guava",
      "currentVersion": "28.0-jre",
      "isSingleVersion": true,
      "fixedVersion": "28.0-jre"
    },
    {
      "depName": "junit:junit",
      "currentValue": "4.12",
      "managerData": {
        "fileReplacePosition": 185,
        "packageFile": "Test.Build.Gradle.Java/build.gradle"
      },
      "fileReplacePosition": 185,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2"
      ],
      "depIndex": 1,
      "updates": [
        {
          "bucket": "non-major",
          "newVersion": "4.13.2",
          "newValue": "4.13.2",
          "newMajor": 4,
          "newMinor": 13,
          "updateType": "minor",
          "branchName": "renovate/all-minor-patch"
        }
      ],
      "warnings": [],
      "versioning": "gradle",
      "sourceUrl": "https://github.com/junit-team/junit4",
      "homepage": "http://junit.org",
      "currentVersion": "4.12",
      "isSingleVersion": true,
      "fixedVersion": "4.12"
    },
    {
      "depName": "org.junit.jupiter:junit-jupiter-engine",
      "currentValue": "5.4.2",
      "managerData": {
        "fileReplacePosition": 252,
        "packageFile": "Test.Build.Gradle.Java/build.gradle"
      },
      "fileReplacePosition": 252,
      "registryUrls": [
        "https://repo.maven.apache.org/maven2"
      ],
      "depIndex": 2,
      "updates": [
        {
          "bucket": "non-major",
          "newVersion": "5.8.1",
          "newValue": "5.8.1",
          "newMajor": 5,
          "newMinor": 8,
          "updateType": "minor",
          "branchName": "renovate/all-minor-patch"
        }
      ],
      "warnings": [],
      "versioning": "gradle",
      "sourceUrl": "https://github.com/junit-team/junit5",
      "homepage": "https://junit.org/junit5/",
      "currentVersion": "5.4.2",
      "isSingleVersion": true,
      "fixedVersion": "5.4.2"
    }
  ],
```

Am I understanding this correctly? Or is there something else going on here?

**Comment by japj** [comment]
(sigh, this is not an answer, still getting used to this discussion mechanism)

I have also seen an occurrence where an PR was created for a build.gradle that intended to update a (private) package version, but instead resulted in:

```
intTestImplementation  ('org.springframework.restdocs:spring-restdocs-mockmvc')

changed into

intTestImplementation  ('org.springf1.0.10:spring-restdocs-mockmvc')
``` 

(note here that 1.0.10 is inserted for the wrong package, at a location that is not related (but does happen to be a package)

**Reply by rarkins** [reply]
Are you applying any special Renovate config, e.g. groupings, or gradle deep mode, etc?

**Reply by japj** [reply]
No custom groupings (just whatever is in config:base) and I believe we were using gradle-lite at some point (not sure if that is the default now). As for gradle deep mode, if that requires explicitly turning it on then we definitely did NOT use that.

**Reply by rarkins** [reply]
Can you reproduce it in a public repo?

**Comment by HonkingGoose** [comment]
@rarkins @viceice Can you take a look at this discussion and see what we need to proceed? This discussion has not seen a response from the maintainers yet. So that's why I'm "nudging" you. 😉

**Comment by linuxhelf** [comment]
I'm having the same issue using renovate 41.42.6

A lot of my dependencies can't be updated with message"Wrong offset" followed by "Error updating file"
```
DEBUG: Wrong offset (repository=privaterepo, branch=renovate/org.json-json-20250517.x)
       "depName": "org.json:json",
       "currentValue": "20240303",
       "newValue": "20250517"
DEBUG: Error updating file (repository=privaterepo, branch=renovate/org.json-json-20250517.x)
       "existingContent": "plugins {\n    id 'application' \n}\n\nrepositories {\n    mavenCentral() \n}\n\ndependencies {\n  implementation 'org.json:json:20240303'\n}\n\napplication {\n    mainClass = 'demo.App' \n}",
       "config": {
```
when modifying the sourcecode of renovate for more information, it seems manager/gradle/update.js expects to get a position where the version is, but the position is wrong.
if i log leftPart + "XXX" + rightPart, i get a position that is behind the version or even somewhere in the next line
But the version to be replaced is expect to be at the beginning of rightPart
e.g.
```
       dependencies {
         implementation 'org.json:json:20240303'XXX
       }
```

I was able to reproduce this in multiple repositories that I use, even in a public repo of mine: (from which I took the log output/examples for this comment)
https://github.com/linuxhelf/vscode-java-import

**Reply by Churro** [reply, marked as answer]
Not reproducible on my end. 

You may want to check this: https://docs.renovatebot.com/getting-started/installing-onboarding/#self-hosting-on-windows

**Reply by linuxhelf** [reply]
hi @Churro 

thanks for the advice.
When I add a few new lines with CRLF the position moves according to the added lines.
Changing the build file to LF seems to fix the parsing.

So my problem is fixed.... but this problem can happen to others users as well
Would it be possible to improve the parsers for Windows users or at least add warnings about CRLF to the debug log?
Or hopefully they will also find this discussion the same as I did

**Reply by Churro** [reply]
The parsers in renovate work fine with all line endings, but the "confusion" arises from the fact that the files in your reproduction repo have LF endings, while your local git client has been explicitly configured with `core.autocrlf=true`. 

I'd argue that renovate shouldn't work around a git setting that is generally discouraged and prominently advised against in renovate's getting-started documentation.
